cassandra-management/src/main/java/org/hawkular/cassandra/management/CassandraMaintenanceJob.java

```diff
@@ -0,0 +1,130 @@
+
+    private ConfigurationService configurationService;
+
+    private EventDataAccess eventDataAccess;
+
+    private RxSession session;
+
+    public CassandraMaintenanceJob(RxSession session, EventDataAccess eventDataAccess,
+            ConfigurationService configurationService) {
+        this.session = session;
+        this.eventDataAccess = eventDataAccess;
+        this.configurationService = configurationService;
+    }
+
+    @Override
+    public Completable call(JobDetails jobDetails) {
+        logger.info("Starting Cassandra cluster maintenance");
+        Stopwatch stopwatch = Stopwatch.createStarted();
+
+        String jmxPort = jobDetails.getParameters().get("jmxPort");
+        String keyspace = jobDetails.getParameters().get("keyspace");
+
+        return eventDataAccess.findEventsForBucket(getCurrentBucket())
+                .collect(ClusterState::new, (state, event) -> {
+                    try {
+                        String hostname = event.getDetails().get("hostname");
+                        InetAddress address = InetAddress.getByName(hostname);
+                        state.setLastNodeAdded(address);
+                    } catch (UnknownHostException e) {
+                        throw new RuntimeException(e);
+                    }
+                })
+                .flatMap(state -> {
+                    if (state.getLastNodeAdded() == null) {
+                        return Observable.empty();
+                    }
+                    return getNodes().filter(address -> !address.equals(state.getLastNodeAdded()));
+                })
+                .concatMap(address -> createExecutionContexts(address, getTables(keyspace), jmxPort, keyspace))
+                .flatMap(context -> new CleanupCommand().run(context))
+                .toCompletable()
+                .concatWith(Completable.defer(() -> configurationService.save(CONFIG_ID, "lastRun",
+                        Long.toString(jobDetails.getTrigger().getTriggerTime())).toCompletable()))
+                .doOnCompleted(() -> {
+                    stopwatch.stop();
+                    logger.infof("Finished Cassandra cluster maintenance in %d ms",
+                            stopwatch.elapsed(TimeUnit.MILLISECONDS));
+                })
+                .doOnError(t -> logger.warn("Cassandra cluster maintenance has been aborted to to an unexpected error",
+                        t));
+    }
+
+    private long getCurrentBucket() {
+        return DateTimeService.current24HourTimeSlice().getMillis();
+    }
+
+    private Observable<InetAddress> getNodes() {
+        return Observable.from(session.getCluster().getMetadata().getAllHosts()).map(Host::getAddress);
+    }
+
+    private Observable<String> getTables(String keyspace) {
+        return Observable.from(session.getCluster().getMetadata().getKeyspace(keyspace).getTables())
+                .map(TableMetadata::getName)
+                .cache();
```

Comment:
Is this cache() in the correct place? The caller does not know some result is cached and this isn't actually cached for multiple callers.